### ISSUE TITLE
Numeric facet with formatting

### ISSUE THREAD

**Issue by jan-hudec** [ogham/rust-locale#17]
``` rust
use std::fmt::{Display,Formatter,Result};

trait Numeric {
    fn fmt_int<N>(n: N, fmt: &str, out: &mut Formatter) -> Result
        where N: Display;
    fn fmt_float<X>(x: X, fmt: &str, out &mut Formatter) -> Result
        where X: Display + LowerExp;
}
```

My plan is to rely on `Display` to get the value formatted to desired precision in invariant digits and then transliterate the symbols and insert group separators. I even have most of that code written already, only patterns are needed.

The `fmt` should be a CLDR-style format string. They are actually pretty hard to parse, but I think we should have them if we are going to have ICU-compatible format specifications if we are going to use ICU-compatible message format.

This facet will _not_ handle formatting monetary values. That will have separate facet, though it will share the formatting code.

Open questions:
- By value, by reference or by Borrow? For plain old numbers, value is appropriate, but this should work for bignums too and those need reference.
- Should we really use the CLDR-style format strings? They are rather confusing in that some of their features need to be interpreted when reading the CLDR data and other features need to be specified from the format string.

**Comment by jan-hudec** [comment]
@waywardmonkeys, do you have experience parsing the numeric format specifiers (like `#,###.##`, `##@@` and such)? I was rather intimidated by them. Arbitrary specifications (say `#+##`) don't really make sense and the restrictions don't seem to be precisely spelled out.

I think for start we should do with just number of decimal digits, number of significant digits and exponential format with one integral digit only. No rounding to multiple or formats with type ("2 thousand" and such — btw. I think at least the long form needs plural forms, but CLDR does not seem to have them).

**Comment by jan-hudec** [comment]
@waywardmonkeys, what does ICU actually use? The patterns clearly need to be interpreted differently when read from CLDR data and when read from the format string and it seems to be rather confusing.

**Comment by waywardmonkeys** [comment]
I don't have any experience with them!

**Comment by jan-hudec** [comment]
@waywardmonkeys, that's pity.

The problem I have with the patterns is that there are aspects that should be taken from the locale (grouping, sign placement, for monetary also precision) and aspects that should be specified at the call site (padding, precision for numbers, but not monetary, international or local symbol for monetary). And the patterns mix them all together. And remember, that I wanted an option for specifying the pattern separately from the format string, because I, generally, do _not_ want to bother translators with them (while still allowing to _override_ them in the translation if needed in some rare case), so the patterns should not specify the aspects taken from the locale. Nor suggest they do.

Hm, I'll have to look what ICU actually uses and then decide how much to follow it and how much to choose something else.

**Comment by jan-hudec** [comment]
So I looked at [ICU documentation](https://ssl.icu-project.org/apiref/icu4c/classicu_1_1MessageFormat.html) and I am **disappointed**. Basically ICU can only use default and percentage format (and integer and monetary, but we are going to select those based on type only), or the CLDR-style pattern, which seems to be the only way to define the precision, but that specifies the grouping too.

I don't consider that sufficient as I believe being able to specify precision is important. So I will proceed using the python3-style specifier of the form “[ _align_ ] [ _sign_ ] [ _fill_ ] [ _width_ ] [ `.` _precision_ ] [ _format-type_ ]” (more-or-less the same as C, but the type is optional too, defaulting to `f` (normal decimal number), probably.

The underlying machinery (of which I have a prototype lying around somewhere) is actually capable of handling minimum and maximum number of decimal digits, but I don't want to overcomplicate the pattern, so the public interface will likely only provide two options: exact precision and maximum precision with trailing zeroes omitted.

**Comment by jan-hudec** [comment]
It might also be worth considering the [.NET composite formatting](https://msdn.microsoft.com/en-us/library/txafckwd%28v=vs.110%29.aspx) for specifying the padding in a general way, but specifically numbers are the odd man out here with their inner padding (between sign and value) and their special behaviour regarding use of `0` as padding character (it uses the locale symbol and has separators inserted), so generic alignment support won't properly work for it.

**Comment by jan-hudec** [comment]
This is now ready on the next branch.
